backend/src/apiserver/client_manager/client_manager.go

```diff
@@ -182,48 +202,88 @@ func (c *ClientManager) Authenticators() []auth.Authenticator {
 	return c.authenticators
 }
 
-func (c *ClientManager) ControllerClient() ctrlclient.Client {
-	return c.controllerClient
-}
+func (c *ClientManager) init(options *Options) error {
+	// time
+	c.time = util.NewRealTime()
 
-func (c *ClientManager) init() error {
-	glog.Info("Initializing controller client...")
+	// UUID generator
+	c.uuid = util.NewUUIDGenerator()
 
-	restConfig, err := util.GetKubernetesConfig()
-	if err != nil {
-		return fmt.Errorf("failed to initialize the RestConfig: %w", err)
-	}
+	var controllerClient ctrlclient.Client
 
-	controllerClient, err := ctrlclient.New(
-		restConfig, ctrlclient.Options{Scheme: scheme},
-	)
-	if err != nil {
-		return fmt.Errorf("failed to initialize the controller client: %w", err)
-	}
-	c.controllerClient = controllerClient
-	glog.Info("Controller client initialized successfully.")
+	var pipelineStoreForRef storage.PipelineStoreInterface
 
-	if common.IsOnlyKubernetesWebhookMode() {
-		return nil
+	if options.UsePipelineKubernetesStorage || common.IsOnlyKubernetesWebhookMode() {
+		glog.Info("Initializing controller client...")
+		restConfig, err := util.GetKubernetesConfig()
+		if err != nil {
+			return err
+		}
+
+		var cacheConfig map[string]cache.Config
+
+		if !common.IsMultiUserMode() && common.GetPodNamespace() != "" && !common.IsOnlyKubernetesWebhookMode() {
+			cacheConfig = map[string]cache.Config{common.GetPodNamespace(): {}}
+		}
+
+		k8sAPICache, err := cache.New(restConfig,
+			cache.Options{
+				DefaultNamespaces: cacheConfig,
+				Scheme:            scheme,
+			},
+		)
+		if err != nil {
+			return err
+		}
+
+		options.WaitGroup.Add(1)
+		go func() {
+			defer options.WaitGroup.Done()
+
+			err := k8sAPICache.Start(options.Context)
+			if err != nil {
+				panic(fmt.Sprintf("Failed to start the cache to the cluster: %v", err))
+			}
+		}()
+
+		controllerClient, err = ctrlclient.New(
+			restConfig, ctrlclient.Options{Scheme: scheme, Cache: &ctrlclient.CacheOptions{Reader: k8sAPICache}},
+		)
+		if err != nil {
+			return fmt.Errorf("failed to initialize the controller client: %w", err)
+		}
+
+		dynamicClient, err := dynamic.NewForConfig(restConfig)
+		if err != nil {
+			return fmt.Errorf("failed to initialize the dynamic client: %w", err)
+		}
+
+		glog.Info("Controller client initialized successfully.")
+
+		c.controllerClient = controllerClient
+		c.dynamicClient = dynamicClient
```

Comment:
why use a dynamic client when we have the known pipeline / pipelineversion type scripts? it seems [we partly add the scheme register init ](https://github.com/kubeflow/pipelines/blob/1e8328960804bf00db2526492799b36a5a4c9d33/backend/src/crd/kubernetes/v2beta1/pipeline_types.go#L97) but could go a step forward and register this scheme in to the package via .AddToScheme, and just use controllerClient for our noCache client as well right? this way we can get [type safety](https://github.com/kubeflow/pipelines/blob/1e8328960804bf00db2526492799b36a5a4c9d33/backend/src/apiserver/storage/pipeline_store_kubernetes.go#L494). Unless there's a special reason to use this.